loopy/type_inference.py

```diff
@@ -393,20 +569,116 @@ def map_reduction(self, expr, return_tuple=False):
                 rec_results = self.rec(expr.expr)
 
         if return_tuple:
-            return [expr.operation.result_dtypes(self.kernel, *rec_result)
+            return [expr.operation.result_dtypes(*rec_result)
                     for rec_result in rec_results]
         else:
-            return [expr.operation.result_dtypes(self.kernel, rec_result)[0]
+            return [expr.operation.result_dtypes(rec_result)[0]
                     for rec_result in rec_results]
 
+    def map_sub_array_ref(self, expr):
+        return self.rec(expr.subscript)
+
+    map_fortran_division = map_quotient
+
+# }}}
+
+
+# {{{ TypeReader
+
+class TypeReader(TypeInferenceMapper):
```

Comment:
If I'm understanding this correctly, then `TypeInferenceMapper` is written to make modifications to callables, while `TypeReader` does not. I'd be more comfortable having the inheritance go the other way.